--- mozilla/browser/app/profile/firefox.js.BAK  2005-10-16 22:33:08.484375000 +0800
+++ mozilla/browser/app/profile/firefox.js      2005-10-16 22:36:42.296875000 +0800
@@ -542,3 +542,45 @@
 pref("browser.sessionstore.privacy_level", 1);
 // number of remembered reopenable tabs (per window)
 pref("browser.sessionstore.max_tabs_undo", 10);
+
+//
+// pigfoot's tweaks (jayfromtaiwan's, MadmanNova's firefox.js)
+//
+pref("browser.enable_automatic_image_resizing", false);
+pref("browser.display.show_image_placeholders", false);
+pref("browser.cache.memory.capacity", 65536);
+pref("browser.cache.disk_cache_ssl", true);
+pref("browser.xul.error_pages.enabled", true);
+pref("browser.turbo.enabled", true);
+//pref("browser.urlbar.autoFill", true);
+//pref("browser.urlbar.autocomplete.learning.mode", 2);
+pref("content.interrupt.parsing", true);
+pref("content.max.tokenizing.time", 3000000);
+pref("content.maxtextrun", 8191);
+pref("content.notify.backoffcount", 5);
+pref("content.notify.interval", 750000);
+pref("content.notify.ontimer", true);
+pref("content.switch.threshold", 750000);
+pref("dom.max_script_run_time", 10);
+pref("dom.popup_maximum", 8);
+pref("dom.disable_window_flip", true);
+pref("dom.disable_window_status_change", true);
+pref("network.http.pipelining", true);
+pref("network.http.proxy.pipelining", true);
+pref("network.http.pipelining.maxrequests", 8);
+pref("network.http.max-connections", 48);
+pref("network.http.max-connections-per-server", 16);
+pref("network.http.max-persistent-connections-per-proxy", 16);
+pref("network.http.max-persistent-connections-per-server", 8);
+pref("nglayout.initialpaint.delay", 500);
+pref("nglayout.initialpaint.delay", 0);
+pref("plugin.expose_full_path", true);
+pref("signed.applets.codebase_principal_support", true);
+pref("ui.submenuDelay", 0);
+pref("config.trim_on_minimize", true);
+
+// Set by pigfoot, For Tranditional Chinese font
+//pref("font.size.fixed.zh-TW", 13);
+pref("general.useragent.extra.pigfoot", "(pigfoot)");
+pref("app.update.url", "https://update.pigfoot.org/__pigfoot_arch__/2/%PRODUCT%/%VERSION%/%BUILD_ID%/%BUILD_TARGET%/%LOCALE%/%CHANNEL%/%OS_VERSION%/update.xml");
+//pref("extensions.checkCompatibility", false);  // Bug330895

--- mozilla/mail/app/profile/all-thunderbird.js.BAK	2005-04-13 00:19:25.171875000 +0800
+++ mozilla/mail/app/profile/all-thunderbird.js	2005-04-13 00:42:39.906250000 +0800
@@ -412,3 +412,10 @@
 
 // prevent status-bar spoofing even if people are foolish enough to turn on JS
 pref("dom.disable_window_status_change",          true);
+
+// Optimization tweaks for a 'speedy' effect by pigfoot
+pref("network.http.pipelining", true);
+pref("network.http.proxy.pipelining", true);
+pref("network.http.pipelining.maxrequests", 100);
+pref("nglayout.initialpaint.delay", 0);
+//expand pref("general.useragent.extra.thunderbird", "Thunderbird/__APP_VERSION__ (pigfoot)");
