--- mozilla/security/nss/lib/ckfw/builtins/certdata.c
+++ mozilla/security/nss/lib/ckfw/builtins/certdata.c
@@ -4640,19 +4640,19 @@ static const NSSItem nss_builtins_items_
 "\145\164\040\114\151\155\151\164\145\144\061\072\060\070\006\003"
 "\125\004\003\023\061\105\156\164\162\165\163\164\056\156\145\164"
 "\040\123\145\143\165\162\145\040\123\145\162\166\145\162\040\103"
 "\145\162\164\151\146\151\143\141\164\151\157\156\040\101\165\164"
 "\150\157\162\151\164\171"
 , (PRUint32)198 },
   { (void *)"\002\004\067\112\322\103"
 , (PRUint32)6 },
-  { (void *)&ckt_netscape_trusted_delegator, (PRUint32)sizeof(CK_TRUST) },
-  { (void *)&ckt_netscape_trusted_delegator, (PRUint32)sizeof(CK_TRUST) },
-  { (void *)&ckt_netscape_trusted_delegator, (PRUint32)sizeof(CK_TRUST) },
+  { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
+  { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
+  { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
   { (void *)&ck_false, (PRUint32)sizeof(CK_BBOOL) }
 };
 static const NSSItem nss_builtins_items_64 [] = {
   { (void *)&cko_certificate, (PRUint32)sizeof(CK_OBJECT_CLASS) },
   { (void *)&ck_true, (PRUint32)sizeof(CK_BBOOL) },
   { (void *)&ck_false, (PRUint32)sizeof(CK_BBOOL) },
   { (void *)&ck_false, (PRUint32)sizeof(CK_BBOOL) },
   { (void *)"Entrust.net Secure Personal CA", (PRUint32)31 },
@@ -19935,17 +19935,17 @@ static const NSSItem nss_builtins_items_
 , (PRUint32)16 },
   { (void *)"\060\062\061\013\060\011\006\003\125\004\006\023\002\103\116\061"
 "\016\060\014\006\003\125\004\012\023\005\103\116\116\111\103\061"
 "\023\060\021\006\003\125\004\003\023\012\103\116\116\111\103\040"
 "\122\117\117\124"
 , (PRUint32)52 },
   { (void *)"\002\004\111\063\000\001"
 , (PRUint32)6 },
-  { (void *)&ckt_netscape_trusted_delegator, (PRUint32)sizeof(CK_TRUST) },
+  { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
   { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
   { (void *)&ckt_netscape_trust_unknown, (PRUint32)sizeof(CK_TRUST) },
   { (void *)&ck_false, (PRUint32)sizeof(CK_BBOOL) }
 };
 static const NSSItem nss_builtins_items_296 [] = {
   { (void *)&cko_certificate, (PRUint32)sizeof(CK_OBJECT_CLASS) },
   { (void *)&ck_true, (PRUint32)sizeof(CK_BBOOL) },
   { (void *)&ck_false, (PRUint32)sizeof(CK_BBOOL) },
--- mozilla/security/nss/lib/ckfw/builtins/certdata.txt
+++ mozilla/security/nss/lib/ckfw/builtins/certdata.txt
@@ -4014,19 +4014,19 @@ CKA_ISSUER MULTILINE_OCTAL
 \125\004\003\023\061\105\156\164\162\165\163\164\056\156\145\164
 \040\123\145\143\165\162\145\040\123\145\162\166\145\162\040\103
 \145\162\164\151\146\151\143\141\164\151\157\156\040\101\165\164
 \150\157\162\151\164\171
 END
 CKA_SERIAL_NUMBER MULTILINE_OCTAL
 \002\004\067\112\322\103
 END
-CKA_TRUST_SERVER_AUTH CK_TRUST CKT_NETSCAPE_TRUSTED_DELEGATOR
-CKA_TRUST_EMAIL_PROTECTION CK_TRUST CKT_NETSCAPE_TRUSTED_DELEGATOR
-CKA_TRUST_CODE_SIGNING CK_TRUST CKT_NETSCAPE_TRUSTED_DELEGATOR
+CKA_TRUST_SERVER_AUTH CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
+CKA_TRUST_EMAIL_PROTECTION CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
+CKA_TRUST_CODE_SIGNING CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
 CKA_TRUST_STEP_UP_APPROVED CK_BBOOL CK_FALSE
 
 #
 # Certificate "Entrust.net Secure Personal CA"
 #
 CKA_CLASS CK_OBJECT_CLASS CKO_CERTIFICATE
 CKA_TOKEN CK_BBOOL CK_TRUE
 CKA_PRIVATE CK_BBOOL CK_FALSE
@@ -20469,17 +20469,17 @@ CKA_ISSUER MULTILINE_OCTAL
 \060\062\061\013\060\011\006\003\125\004\006\023\002\103\116\061
 \016\060\014\006\003\125\004\012\023\005\103\116\116\111\103\061
 \023\060\021\006\003\125\004\003\023\012\103\116\116\111\103\040
 \122\117\117\124
 END
 CKA_SERIAL_NUMBER MULTILINE_OCTAL
 \002\004\111\063\000\001
 END
-CKA_TRUST_SERVER_AUTH CK_TRUST CKT_NETSCAPE_TRUSTED_DELEGATOR
+CKA_TRUST_SERVER_AUTH CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
 CKA_TRUST_EMAIL_PROTECTION CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
 CKA_TRUST_CODE_SIGNING CK_TRUST CKT_NETSCAPE_TRUST_UNKNOWN
 CKA_TRUST_STEP_UP_APPROVED CK_BBOOL CK_FALSE
 
 #
 # Certificate "ApplicationCA - Japanese Government"
 #
 CKA_CLASS CK_OBJECT_CLASS CKO_CERTIFICATE
