--- l10n/fr/other-licenses/branding/firefox/brand.dtd.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/fr/other-licenses/branding/firefox/brand.dtd	Tue Feb  3 23:23:41 2009
@@ -1,3 +1,3 @@
 <!ENTITY  brandShortName        "Firefox">
-<!ENTITY  brandFullName         "Mozilla Firefox">
+<!ENTITY  brandFullName         "Firefox">
 <!ENTITY  vendorShortName       "Mozilla">
--- l10n/fr/other-licenses/branding/firefox/brand.properties.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/fr/other-licenses/branding/firefox/brand.properties	Tue Feb  3 23:23:41 2009
@@ -1,5 +1,5 @@
 brandShortName=Firefox
-brandFullName=Mozilla Firefox
+brandFullName=Firefox
 vendorShortName=Mozilla
 
 homePageSingleStartMain=Firefox Start, une page d'accueil rapide avec recherche intégrée
--- l10n/de/other-licenses/branding/firefox/brand.dtd.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/de/other-licenses/branding/firefox/brand.dtd	Tue Feb  3 23:23:41 2009
@@ -1,3 +1,3 @@
 <!ENTITY  brandShortName        "Firefox">
-<!ENTITY  brandFullName         "Mozilla Firefox">            
+<!ENTITY  brandFullName         "Firefox">            
 <!ENTITY  vendorShortName       "Mozilla">
--- l10n/de/other-licenses/branding/firefox/brand.properties.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/de/other-licenses/branding/firefox/brand.properties	Tue Feb  3 23:23:41 2009
@@ -1,5 +1,5 @@
 brandShortName=Firefox
-brandFullName=Mozilla Firefox
+brandFullName=Firefox
 vendorShortName=Mozilla
 
 homePageSingleStartMain=Firefox Start, eine schnelle Startseite mit integrierter Suche
--- l10n/ja/other-licenses/branding/firefox/brand.dtd.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/ja/other-licenses/branding/firefox/brand.dtd	Tue Feb  3 23:23:41 2009
@@ -1,3 +1,3 @@
 <!ENTITY  brandShortName	"Firefox">
-<!ENTITY  brandFullName		"Mozilla Firefox">
+<!ENTITY  brandFullName		"Firefox">
 <!ENTITY  vendorShortName	"Mozilla">
--- l10n/ja/other-licenses/branding/firefox/brand.properties.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/ja/other-licenses/branding/firefox/brand.properties	Tue Feb  3 23:23:41 2009
@@ -1,5 +1,5 @@
 brandShortName			= Firefox
-brandFullName			= Mozilla Firefox
+brandFullName			= Firefox
 vendorShortName			= Mozilla
 
 homePageSingleStartMain		= Firefox Start - 組み込みの検索エンジンによるスタートページ
--- l10n/zh-CN/other-licenses/branding/firefox/brand.dtd.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/zh-CN/other-licenses/branding/firefox/brand.dtd	Tue Feb  3 23:23:41 2009
@@ -1,3 +1,3 @@
 <!ENTITY  brandShortName        "Firefox">
-<!ENTITY  brandFullName         "Mozilla Firefox">
+<!ENTITY  brandFullName         "Firefox">
 <!ENTITY  vendorShortName       "Mozilla">
--- l10n/zh-CN/other-licenses/branding/firefox/brand.properties.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/zh-CN/other-licenses/branding/firefox/brand.properties	Tue Feb  3 23:23:41 2009
@@ -1,5 +1,5 @@
 brandShortName = Firefox
-brandFullName = Mozilla Firefox
+brandFullName = Firefox
 vendorShortName = Mozilla
 homePageSingleStartMain = Firefox 起始页，内建的快速搜索主页
 homePageImport = 从 %S 导入您的主页
--- l10n/zh-TW/other-licenses/branding/firefox/brand.dtd.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/zh-TW/other-licenses/branding/firefox/brand.dtd	Tue Feb  3 23:23:41 2009
@@ -1,3 +1,3 @@
 <!ENTITY  brandShortName        "Firefox">
-<!ENTITY  brandFullName         "Mozilla Firefox">
+<!ENTITY  brandFullName         "Firefox">
 <!ENTITY  vendorShortName       "Mozilla">
--- l10n/zh-TW/other-licenses/branding/firefox/brand.properties.BAK	Sat Oct 11 03:05:34 2008
+++ l10n/zh-TW/other-licenses/branding/firefox/brand.properties	Tue Feb  3 23:23:41 2009
@@ -1,5 +1,5 @@
 brandShortName=Firefox
-brandFullName=Mozilla Firefox
+brandFullName=Firefox
 vendorShortName=Mozilla
 
 homePageSingleStartMain=Firefox Start，結合快速搜尋的首頁
--- l10n/it/other-licenses/branding/firefox/brand.dtd.BAK	2010-01-22 18:31:29 +0800
+++ l10n/it/other-licenses/branding/firefox/brand.dtd	2010-01-22 19:41:47 +0800
 <!ENTITY brandShortName "Firefox">
-<!ENTITY brandFullName "Mozilla Firefox">
+<!ENTITY brandFullName "Firefox">
 <!ENTITY vendorShortName "Mozilla">
 <!ENTITY logoCopyright "Firefox e i logo di Firefox sono marchi registrati della Mozilla Foundation. Tutti i diritti riservati.">
--- l10n/it/other-licenses/branding/firefox/brand.properties.BAK	2010-01-22 18:25:00 +0800
+++ l10n/it/other-licenses/branding/firefox/brand.properties	2010-01-22 18:31:40 +0800
@@ -1,5 +1,5 @@
 brandShortName = Firefox
-brandFullName = Mozilla Firefox
+brandFullName = Firefox
 vendorShortName = Mozilla
 homePageSingleStartMain = Pagina iniziale di Firefox, leggera e con ricerca integrata
 homePageImport = Importa la pagina iniziale da %S
--- l10n/ru/other-licenses/branding/firefox/brand.dtd.BAK	2010-01-22 18:25:30 +0800
+++ l10n/ru/other-licenses/branding/firefox/brand.dtd	2010-01-22 18:32:52 +0800
@@ -1,4 +1,4 @@
 <!ENTITY brandShortName "Firefox">
-<!ENTITY brandFullName "Mozilla Firefox">
+<!ENTITY brandFullName "Firefox">
 <!ENTITY vendorShortName "Mozilla">
 <!ENTITY logoCopyright "Firefox и логотипы Firefox являются торговыми марками Mozilla Foundation. Все права защищены.">
--- l10n/ru/other-licenses/branding/firefox/brand.properties.BAK	2010-01-22 18:25:30 +0800
+++ l10n/ru/other-licenses/branding/firefox/brand.properties	2010-01-22 18:32:48 +0800
@@ -1,5 +1,5 @@
 brandShortName = Firefox
-brandFullName = Mozilla Firefox
+brandFullName = Firefox
 vendorShortName = Mozilla
 homePageSingleStartMain = Начальная страница Firefox — быстрая, со встроенным поиском
 homePageImport = Импортировать домашнюю страницу из %S
--- l10n/sl/other-licenses/branding/firefox/brand.dtd.BAK	2010-06-25 14:34:19 +0800
+++ l10n/sl/other-licenses/branding/firefox/brand.dtd	2010-06-25 14:36:07 +0800
@@ -1,4 +1,4 @@
 <!ENTITY brandShortName "Firefox">
-<!ENTITY brandFullName "Mozilla Firefox">
+<!ENTITY brandFullName "Firefox">
 <!ENTITY vendorShortName "Mozilla">
 <!ENTITY  logoCopyright         "Firefox in logotip Firefoxa sta blagovni znamki Mozilla Foundation. Vse pravice pridržane.">
--- l10n/sl/other-licenses/branding/firefox/brand.properties.BAK	2010-06-25 14:34:19 +0800
+++ l10n/sl/other-licenses/branding/firefox/brand.properties	2010-06-25 14:36:15 +0800
@@ -1,5 +1,5 @@
 brandShortName=Firefox
-brandFullName=Mozilla Firefox
+brandFullName=Firefox
 vendorShortName=Mozilla
 
 homePageSingleStartMain=Firefox Start, hitra domača stran z vgrajenim iskalnikom
--- l10n/tr/other-licenses/branding/firefox/brand.dtd.BAK	2010-10-21 18:12:48 +0800
+++ l10n/tr/other-licenses/branding/firefox/brand.dtd.	2010-10-21 18:19:55 +0800
@@ -1,4 +1,4 @@
 <!ENTITY brandShortName "Firefox">
-<!ENTITY brandFullName "Mozilla Firefox">
+<!ENTITY brandFullName "Firefox">
 <!ENTITY vendorShortName "Mozilla">
 <!ENTITY logoCopyright "Firefox ve Firefox logoları Mozilla Vakfı'nın kayıtlı markalarıdır. Her hakkı saklıdır.">
--- l10n/tr/other-licenses/branding/firefox/brand.properties.BAK	2010-10-21 18:12:48 +0800
+++ l10n/tr/other-licenses/branding/firefox/brand.properties	2010-10-21 18:20:02 +0800
@@ -1,5 +1,5 @@
 brandShortName=Firefox
-brandFullName=Mozilla Firefox
+brandFullName=Firefox
 vendorShortName=Mozilla
 
 homePageSingleStartMain=Yerleşik arama özelliğiyle Firefox Başlangıç Sayfası
