Index: gfx/public/nsCoord.h
===================================================================
RCS file: /cvsroot/mozilla/gfx/public/nsCoord.h,v
retrieving revision 1.16
diff -u -8 -p -r1.16 nsCoord.h
--- gfx/public/nsCoord.h	8 Apr 2008 00:27:10 -0000	1.16
+++ gfx/public/nsCoord.h	10 Aug 2008 02:34:28 -0000
@@ -287,17 +287,19 @@ inline nscoord NSToCoordFloor(float aVal
 
 inline nscoord NSToCoordCeil(float aValue)
 {
   return nscoord(NS_ceilf(aValue));
 }
 
 inline nscoord NSToCoordRound(float aValue)
 {
-#if defined(XP_WIN32) && defined(_M_IX86) && !defined(__GNUC__)
+#if defined(TT_MEMUTIL) && _M_IX86_FP >= 1
+  return ((0.0f <= aValue) ? nscoord(aValue + 0.5f) : nscoord(aValue - 0.5f));
+#elif defined(XP_WIN32) && defined(_M_IX86) && !defined(__GNUC__)
   return NS_lroundup30(aValue);
 #else
   return nscoord(NS_floorf(aValue + 0.5f));
 #endif /* XP_WIN32 && _M_IX86 && !__GNUC__ */
 }
 
 /*
  * Int Rounding Functions
