Index: xpcom/glue/nsDeque.cpp
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsDeque.cpp,v
retrieving revision 1.1
diff -u -8 -p -r1.1 nsDeque.cpp
--- xpcom/glue/nsDeque.cpp	18 Sep 2007 18:14:35 -0000	1.1
+++ xpcom/glue/nsDeque.cpp	7 Feb 2008 13:50:27 -0000
@@ -137,17 +137,32 @@ void nsDeque::SetDeallocator(nsDequeFunc
 
 /**
  * Remove all items from container without destroying them.
  *
  * @return  *this
  */
 nsDeque& nsDeque::Empty() {
   if (mSize && mData) {
+#ifdef TT_MEMUTIL
+    static const PRUint32 dwNonTemporalDataSizeMin = GetNonTemporalDataSizeMin_tt();
+    const PRUint32 dwDataSize = mCapacity*sizeof(mData);
+
+    if (dwDataSize < dwNonTemporalDataSizeMin ||
+        NON_TEMPORAL_STORES_NOT_SUPPORTED == dwNonTemporalDataSizeMin)
+    {
+        memset(mData, 0, mCapacity*sizeof(mData));
+    }
+    else
+    {
+        memset_nontemporal_tt(mData, 0, mCapacity*sizeof(mData));
+    }
+#else
     memset(mData, 0, mCapacity*sizeof(mData));
+#endif
   }
   mSize=0;
   mOrigin=0;
   return *this;
 }
 
 /**
  * Remove and delete all items from container
