Index: xpcom/glue/nsID.h
===================================================================
RCS file: /cvsroot/mozilla/xpcom/glue/nsID.h,v
retrieving revision 1.30
diff -u -8 -p -r1.30 nsID.h
--- xpcom/glue/nsID.h	15 Jan 2008 16:46:15 -0000	1.30
+++ xpcom/glue/nsID.h	13 Feb 2008 08:27:09 -0000
@@ -76,21 +76,18 @@ struct nsID {
   inline PRBool Equals(const nsID& other) const {
     // One would think that this could be done faster with a really
     // efficient implementation of memcmp(), but evidently no
     // memcmp()'s out there are better than this code.
     //
     // See bug http://bugzilla.mozilla.org/show_bug.cgi?id=164580 for
     // details.
 
-    return
-      ((((PRUint32*) &m0)[0] == ((PRUint32*) &other.m0)[0]) &&
-       (((PRUint32*) &m0)[1] == ((PRUint32*) &other.m0)[1]) &&
-       (((PRUint32*) &m0)[2] == ((PRUint32*) &other.m0)[2]) &&
-       (((PRUint32*) &m0)[3] == ((PRUint32*) &other.m0)[3]));
+    return (((PRUint32*) &m0)[0] + ((PRUint32*) &m0)[3]) ==
+        (((PRUint32*) &other.m0)[0] + ((PRUint32*) &other.m0)[3]);
   }
 
   /**
    * nsID Parsing method. Turns a {xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}
    * string into an nsID
    */
   NS_COM_GLUE PRBool Parse(const char *aIDStr);
 
