Index: nsprpub/configure.in
===================================================================
RCS file: /cvsroot/mozilla/nsprpub/configure.in,v
retrieving revision 1.233
diff -u -8 -p -r1.233 configure.in
--- nsprpub/configure.in	28 Apr 2008 23:21:11 -0000	1.233
+++ nsprpub/configure.in	12 May 2008 10:11:04 -0000
@@ -1550,16 +1550,17 @@ tools are selected during the Xcode/Deve
                 MT=mt
                 unset MSMT_TOOL
             else
                 AC_MSG_ERROR([Microsoft (R) Manifest Tool must be in your \$PATH.])
             fi
         fi
         
         CFLAGS="$CFLAGS -W3 -nologo -GF -Gy"
+        DLLFLAGS="$DLLFLAGS $LDFLAGS"
         DLLFLAGS="$DLLFLAGS -OUT:\"\$@\""
         _DEBUG_FLAGS=-Z7
         _OPTIMIZE_FLAGS=-O2
         if test -z "$MOZ_OPTIMIZE"; then
             CFLAGS="$CFLAGS -Od"
         fi
 
         if test -n "$USE_DEBUG_RTL"; then
