Index: layout/style/nsStyleUtil.cpp
===================================================================
RCS file: /cvsroot/mozilla/layout/style/nsStyleUtil.cpp,v
retrieving revision 3.65
diff -u -8 -p -r3.65 nsStyleUtil.cpp
--- layout/style/nsStyleUtil.cpp	1 Jan 2008 10:11:05 -0000	3.65
+++ layout/style/nsStyleUtil.cpp	13 Feb 2008 08:39:07 -0000
@@ -51,17 +51,17 @@
 #include "nsReadableUtils.h"
 #include "nsContentUtils.h"
 #include "nsTextFormatter.h"
 
 // XXX This is here because nsCachedStyleData is accessed outside of
 // the content module; e.g., by nsCSSFrameConstructor.
 #include "nsRuleNode.h"
 
-nsCachedStyleData::StyleStructInfo
+const nsCachedStyleData::StyleStructInfo
 nsCachedStyleData::gInfo[] = {
 
 #define STYLE_STRUCT_INHERITED(name, checkdata_cb, ctor_args) \
   { offsetof(nsCachedStyleData, mInheritedData), \
     offsetof(nsInheritedStyleData, m##name##Data), \
     PR_FALSE },
 #define STYLE_STRUCT_RESET(name, checkdata_cb, ctor_args) \
   { offsetof(nsCachedStyleData, mResetData), \
